ast/src/parsed/mod.rs

```diff
@@ -61,18 +66,192 @@ pub enum Expression<T> {
     ),
 }
 
+impl<T> From<ShiftedPolynomialReference<T>> for Expression<T> {
+    fn from(value: ShiftedPolynomialReference<T>) -> Self {
+        Self::PolynomialReference(value)
+    }
+}
+
 #[derive(Debug, PartialEq, Eq, Default, Clone)]
 pub struct PolynomialName<T> {
     pub name: String,
     pub array_size: Option<Expression<T>>,
 }
 
+/// A polynomial with an optional shift
 #[derive(Debug, PartialEq, Eq, Default, Clone)]
-pub struct PolynomialReference<T> {
-    pub namespace: Option<String>,
-    pub name: String,
-    pub index: Option<Box<Expression<T>>>,
-    pub next: bool,
+pub struct ShiftedPolynomialReference<T> {
+    /// Whether we shift or not
+    is_next: bool,
+    /// The underlying polynomial
+    pol: NamespacedPolynomialReference<T>,
+}
+
+impl<T> ShiftedPolynomialReference<T> {
+    /// Returns the underlying namespaced polynomial
+    pub fn into_namespaced(self) -> NamespacedPolynomialReference<T> {
+        self.pol
+    }
+
+    /// Returns the shift of this polynomial
+    pub fn shift(&self) -> bool {
```

Comment:
Nit: `shift` as a function name sounds like an action to me. I'd suggest `is_shifted` or `is_next` or similar